firmware/STN1110.c

```diff
@@ -163,25 +256,101 @@ void _process_pid_response(char * buf)
         for (i = 0; i < count; i++) {
             can_pid_response.data8[i + 1] = pid_response[i];
         }
-        /* Pause before transmitting the message
+
+        /*mark when we've received the message for our metrics */
+        mark_stn1110_rx();
+
+        /* once we successfully get a PID response, ask
+         * for the detected protocol if we still don't know what it is.
+         * The response will be picked up the next time around in this
+         * function.
+         */
+        if (get_detected_protocol() == obdii_protocol_auto) {
+            _send_detect_protocol();
+        }
+
+        /* Pause before transmitting the message to limit update rate
          * since the other system may immediately send the next PID request
          */
-        chThdSleepMilliseconds(OBDII_PID_POLL_DELAY);
+        chThdSleepMilliseconds(get_pid_poll_delay());
+
+        /* Now send the OBDII response */
         canTransmit(&CAND1, CAN_ANY_MAILBOX, &can_pid_response, MS2ST(CAN_TRANSMIT_TIMEOUT));
+        log_CAN_tx_message(LOG_PFX, &can_pid_response);
+
+        /* We got a response to the OBDII query */
+        got_obd2_response = true;
+
+        /* We've successfully received at least one message;
+         *set the timeout to the runtime timeout
+         */
+        set_obdii_request_timeout(OBDII_RUNTIME_TIMEOUT);
+
+        /* We successfully got a PID response */
+        set_stn1110_error(STN1110_ERROR_NONE);
+
+        /* Reset our NODATA error count since we've successfully received a PID response */
+        reset_nodata_error_count();
+    }
+
+    /* If we got some sort of response - error or success
+     * then mark ourselves as done.
+     */
+    if (got_obd2_response) {
+        log_trace(LOG_PFX "STN1110 latency: %ims\r\n", get_stn1110_latency());
         set_pid_request_active(false);
-        log_info(LOG_PFX "CAN Tx\r\n");
-        _log_CAN_tx_message(&can_pid_response);
+        reset_obdii_timeout_count();
+    }
+}
+
+void send_stn1110_pid_request(uint8_t * data, size_t data_len)
+{
+    /* check if we're in the middle of a PID request,
+     * and if so, did we time out? */
+    if (get_pid_request_active()) {
+        if (is_pid_request_timeout(get_obdii_request_timeout())) {
+            log_info(LOG_PFX "Previous PID request timed out\r\n");
+            increment_obdii_timeout_count();
+            if (get_obdii_timeout_count() > MAX_OBDII_TIMEOUTS) {
+                log_info(LOG_PFX "Max timeouts, resetting system\r\n");
+                reset_system();
```

Comment:
General coding feedback: Putting these sorts of resets buried in methods that are meant for other things (like sending a PID) is really a bad idea because you are breaking the concept of isolation of responsibility of the method.  Instead of this method sending just a PID, it can also now reset the system or ignore timeouts.  This is the road to unmaintainable code.  Instead consider using return variable to indicate whether or not a PID was able to be sent.  This way then this method can be responsible for doing one thing only, keeping things simple and elegant.  The logic can be in some higher method whose purpose is to send or reset.